File: .agents/plugins/opencode-aidevops/structured-text-parser.mjs
// SPDX-License-Identifier: MIT
// SPDX-FileCopyrightText: 2025-2026 Marcus Quinn

/** Parse complete JSON containers without treating scalars as structured output. */
function parseJsonContainer(text) {
  try {
    const value = JSON.parse(text);
    return value !== null && typeof value === "object" ? value : null;
  } catch {
    return null;
  }
}

/** Parse complete JSON or independent NDJSON records while preserving separators. */
export function parseStructuredText(value) {
  const document = parseJsonContainer(value.trim());
  if (document !== null) {
    return { value: document, stringify: (scrubbed) => JSON.stringify(scrubbed) };
  }

  const parts = value.split(/(\r?\n)/);
  const records = [];
  let recordCount = 0;
  for (let index = 0; index < parts.length; index += 2) {
    const line = parts[index];
    if (!line.trim()) {
      records.push(null);
      continue;
    }
    const record = parseJsonContainer(line);
    if (record === null) return null;
    records.push(record);
    recordCount++;
  }
  if (recordCount === 0) return null;

  return {
    value: records,
    stringify: (scrubbed) =>
      scrubbed
        .map((record, index) => {
          const line = record === null ? parts[index * 2] : JSON.stringify(record);
          return `${line}${parts[index * 2 + 1] ?? ""}`;
        })
        .join(""),
  };
}

File: .agents/hooks/structured_text_parser.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
# SPDX-FileCopyrightText: 2025-2026 Marcus Quinn
"""Structured text parsing for transcript scrub hooks."""

import json


def parse_structured_text(value: str):
    """Parse complete JSON or independent NDJSON records for recursive scrubbing."""
    parsed = _parse_json_document(value)
    if parsed is not None:
        return parsed, lambda scrubbed: json.dumps(scrubbed, separators=(",", ":"))
    records = []
    for line in value.splitlines():
        record = _parse_json_document(line) if line.strip() else None
        records.append(line if record is None and line.strip() else record)
    if not any(isinstance(record, (dict, list)) for record in records):
        return None
    return records, lambda scrubbed: "\n".join(_stringify_record(record) for record in scrubbed)


def _parse_json_document(value: str):
    """Return JSON containers and leave scalar or invalid values unparsed."""
    try:
        parsed = json.loads(value)
    except json.JSONDecodeError:
        return None
    return parsed if isinstance(parsed, (dict, list)) else None


def _stringify_record(record) -> str:
    """Serialize one parsed record or preserve one raw NDJSON line."""
    if record is None:
        return ""
    if isinstance(record, str):
        return record
    return json.dumps(record, separators=(",", ":"))
